network-rfc/11_portal_api.md

```diff
@@ -0,0 +1,157 @@
+- `Content-Type: application/json`
+
+The client **may** pass the following request headers:
+- `Accept-Encoding`: one of `gzip` or `deflate`. If omitted, the response will be decompressed by the nginx making streaming less efficient. Once implemented, the `deflate` option should be preferred to avoid decompression.
+
+### Response
+The response is a stream of JSON objects in [JSON Lines](https://jsonlines.org/) format, each representing a block. The response may be empty (with `204` status code) indicating that no blocks are available for the given range yet.
+
+Upon receiving the request, the portal starts streaming the blocks from `fromBlock` in the order of increasing block numbers. The portal may return any number of blocks before terminating the stream. Due to the streaming nature of the HTTP protocol, it's not possible to communicate the reason of termination to the client. Whenever a successful non-empty response has been received, the client should check the last returned block number and send a follow-up request if more blocks are needed.
+
+Aside from the blocks explicitly requested by the query, the response always contains the first and the last block of the scanned range. It may also contain additional blocks not requested by the client.
+
+If a block `N` was included in the response, it is guaranteed that all the existing blocks on the simple path (in terms of block chain) from the `fromBlock` (from the query) to `N` matching the query are also included in the response.
+
+### Errors
+| Reason | HTTP Code | Details |
+|-|-|-|
+| Internal error (crash) during request handling | `500` | The request should not be retried because it may be causing the error |
+| The data chunk exists, but no worker is ready to serve it | `503` | The request should be retried later |
+| All the workers rate limited this portal | `429` | Should be retried after the specified delay |
+| The query is invalid | `400` | The request should not be retried |
+| The dataset is not found | `404` | The client should only request datasets from the list returned by `/datasets` |
+| Invalid endpoint | `404` | |
+| The `fromBlock` is past the last known block | `204` | Those blocks may be present later. It's up to the client to decide how often to check for updates. |
+| The requested block range falls into the dataset range, but doesn't contain any blocks | `200` | Some chains may have gaps in the numbering. The request should not be retried in this case — no blocks will ever appear for such range. |
+| The first requested block is less than the beginning of the dataset range | `400` | The client should make sure the requested range is within the dataset range |
+| The `parentBlockHash` in the request doesn't match the current parent hash for the given `fromBlock` | `409` | See below |
+
+In case of the `503` and `429` codes there may be an additional `Retry-After` header indicating the number of seconds to wait before retrying the request.
+
+### Real-time mode
```

Comment:
No! 

Usage modes were not intended, and there is no difference  between how finalized and unfinalized blocks are handled.

The only difference between finalized and unfinalized part, is that `409` on finalized part means misconfiguration (or entire chain reset), and on unfinalized part it can also mean a "fork".

`/finalized-stream` was introduced to provide a stable API until arrival of hot blocks (and corresponding API correction). 

If no one uses it - it should be deprecated.

For clients, that intend to deal with finalized blocks only, a special option should be provided, but the fetch mechanics should stay the same.